KAFKA-5627; Reduce classes needed for LeaderAndIsrPartitionState and MetadataPartitionState

lindong28 opened the pull request.

lindong28 commented:
@junrao @becketqin This is one of the followup patch for KIP-112.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6289/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6273/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6274/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6290/
Test FAILed (JDK 7 and Scala 2.11).

ijuma commented:
Thanks for the PR. Is this an improvement? It seems like it added more lines that it removed:

+204 −189

Part of that is due to the copyright notice though.

lindong28 commented:
@ijuma Thanks for the quick review. There are a bit more lines because this patch added two new files for BasePartitionState and UpdateMetadataState. Thus 32 lines are added simply due to license.

This is an improvement because it removes two classes in `LeaderAndIsr.scala` and some conversion is no longer needed.

ijuma commented:
@lindong28, can you elaborate on why you made the classes top level? Also, it's not clear why we need the common base class since we don't seem to ever reply on subtyping. It seems to be mostly about implementation inheritance (i.e. save a bit of typing), which is a bit of an anti-pattern.

lindong28 commented:
@ijuma I don't have a strong opinion on this. @becketqin think it is better to make these classes top level and add a BasePartitionState to hold the common variables. Do you prefer not to have this BasePartitionState?

I think MetadataPartitionState is probably better to be a top level class instead of a public static class in UpdateMetadataRequest.java. This is because MetadataCache.scala uses this class directly without having anything to do with UpdateMetadataRequest. What do you think?

becketqin commented:
@ijuma Before KIP-112 the UpdateMetadataRequest and LeaderAndIsrRequest contain the same partition state info. So they basically share the same top level PartitionState class. However, after KIP-112, UpdateMetadataRequest and LeaderAndIsrRequest now has slightly different fields in the partition state. UpdateMetadatRequest has offline replicas list while LeaderAndIsrRequest has isNew. The partition state in UpdateMetadataRequest is also used to maintain controller partition state out of the context of UpdateMetadataRequest. When reviewing #2929 I feel those almost identical but slightly different classes makes the logic less clean. So I suggested to make it clear that we have a common partition state class while the partition state in LeaderAndIsrRequest and UpdateMetadataRequest can just extend from that. This gives a clearer hierarchy. 

I agree that the partition state of a particular request should be put into the corresponding request instead of being top level class. One thing to notice is that the partition state for UpdateMetadataRequest is also used in the metadata cache which has nothing to do with UpdateMetadataRequest itself. So previously this class was a top level scala class but resides in the same scala file with LeaderAndIsr. When we make those class top level java class, a new java file will be required.

So my suggestion is the following:
1. Have a top level CommonPartitionState class
2. Have a nested LeaderAndIsrRequest.PartitonState extends CommonPartitionState
3. Have a top level PartitionState class extends CommonPartitionState. It is used by both Metadata and UpdateMetadataRequest.

What do you think?

ijuma commented:
Thanks for elaborating @lindong28 and @becketqin.

With regards to having a common base class simply for implementation inheritance (i.e. we want to have some common fields, but we never rely on on polymorphism), that's generally discouraged. Composition is recommended instead (i.e. have a class with the common fields and have a field for it in LeaderIsrRequest.PartitionState). Would that work here?

With regards to needing a top level class because it's used in the MetadataCache, not sure I agree. MetadataCache is updated via:

```scala
def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest): Seq[TopicPartition] = {
```

So, it's receiving `UpdataMetadataCache` and caching some of the data. So, it's already coupled with the request and it seems unnecessary to try and hide this. Also, there isn't much of a decoupling given that the top level request class would still live in the requests package and is used in the exact same way by `UpdateMetadataRequest`. What is currently in trunk is actually more decoupled, but it involves some duplication as you both pointed out.

Does that make sense?

becketqin commented:
@ijuma Yes, composition is probably better. I think it is fine to just use the UpdateMetadataRequest.PartitionState in the metadata cache. I was under the impression that Controller was also using that class for internal state management. But it seems just using LeaderAndIsr (which probably should also be migrated to Java code).

lindong28 committed:
KAFKA-5627; Reduce classes needed for LeaderAndIsrPartitionState and MetadataPartitionState

lindong28 commented:
@ijuma @becketqin Thanks much for detailed explanation. I have updated the patch as we discussed. Can you take a look when you get time?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6275/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6291/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6293/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6277/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6294/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6278/
Test PASSed (JDK 8 and Scala 2.12).

lindong28 commented:
@ijuma @becketqin @junrao Here is what this patch does:

- Removed `MetadataPartitionState` and `LeaderAndIsrPartitionState` in LeaderAndIsr.scala
- Moved `org.apache.kafka.common.requests.PartitionState` to `org.apache.kafka.common.requests.LeaderAndIsrRequest.PartitionState` in LeaderAndIsrRequest.java
- Created `BasePartitionState` which contains the fields shared between `LeaderAndIsrRequest.PartitionState` and `UpdateMetadataRequest.PartitionState`
- Both `LeaderAndIsrRequest.PartitionState` and `UpdateMetadataRequest.PartitionState` will have `BasePartitionState` as a field to hold the fields that are shared between them.

becketqin commented on clients/src/main/java/org/apache/kafka/common/requests/BasePartitionState.java:
```diff
@@ -16,38 +16,24 @@
-import org.apache.kafka.common.utils.Utils;
-
 import java.util.List;
 
```
Can we add some comments for this class?

lindong28 replied:
@becketqin Sure. I added this comment. Is this OK?

`This class contains the common fields shared between LeaderAndIsrRequest.PartitionState and UpdateMetadataRequest.PartitionState`.

lindong28 committed:
Add comment

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6316/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6297/
Test PASSed (JDK 8 and Scala 2.12).

lindong28 committed:
Fix indentation

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6327/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6312/
Test PASSed (JDK 8 and Scala 2.12).

lindong28 committed:
Address comment

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6334/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6319/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6321/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6336/
Test PASSed (JDK 7 and Scala 2.11).

becketqin commented:
Thanks for the patch. LGTM.
